float64[3] eta
float64[3] epsilon